### PR TITLE
ENH: Make the id property in to_json optional

### PR DESCRIPTION
**Pull request by jdmcbr**
As discussed in #390, the id property in geojson generated with to_json
may not be all that meaningful (frequently just an arbitrary row number
in a dataframe). The option to disable writing ids without meaning is
given here with a new argument `drop_id`, set to `False` by default.

### CODE REVIEW

**Comment by jdmcbr** [comment]
Looks like the failures here (https://github.com/geopandas/geopandas/pull/1637/checks?check_run_id=1171118616) are also coming up in the totally unrelated pr #1638.

Env in failures:
```
# packages in environment at /usr/share/miniconda/envs/test:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main  
apipkg                    1.5                      py36_0  
arrow-cpp                 0.15.1           py36hfa00764_6  
attrs                     20.2.0                     py_0  
blas                      1.0                         mkl  
boost-cpp                 1.67.0               h14c3975_4  
brotli                    1.0.9                he6710b0_0  
bzip2                     1.0.8                h7b6447c_0  
c-ares                    1.16.1               h7b6447c_0  
ca-certificates           2020.7.22                     0  
cairo                     1.14.12              h8948797_3  
certifi                   2020.6.20                py36_0  
cfitsio                   3.470                hb7c8383_2  
click                     7.1.2                      py_0  
click-plugins             1.1.1                      py_0  
cligj                     0.5.0                    py36_0  
coverage                  5.3              py36h7b6447c_0  
curl                      7.67.0               hbc83047_0  
cycler                    0.10.0                   py36_0  
dbus                      1.13.16              hb2f20db_0  
deprecated                1.2.10                   pypi_0    pypi
descartes                 1.1.0                      py_4  
double-conversion         3.1.5                he6710b0_1  
execnet                   1.7.1                      py_0  
expat                     2.2.9                he6710b0_2  
fiona                     1.8.13.post1     py36hc820daa_0  
fontconfig                2.13.0               h9420a91_0  
freetype                  2.10.2               h5ab3b9f_0  
freexl                    1.0.5                h14c3975_0  
fsspec                    0.8.0                      py_0  
gdal                      3.0.2            py36hbb2a789_0  
geographiclib             1.50                     pypi_0    pypi
geopy                     2.0.0                    pypi_0    pypi
geos                      3.8.0                he6710b0_0  
geotiff                   1.5.1                h21e8280_1  
gflags                    2.2.2                he6710b0_0  
giflib                    5.1.4                h14c3975_1  
glib                      2.63.1               h5a9c865_0  
glog                      0.4.0                he6710b0_0  
grpc-cpp                  1.26.0               hf8bcb03_0  
gst-plugins-base          1.14.0               hbbd80ab_1  
gstreamer                 1.14.0               hb453b48_1  
hdf4                      4.2.13               h3ca952b_2  
hdf5                      1.10.4               hb1b8bf9_0  
icu                       58.2                 he6710b0_3  
importlib-metadata        1.7.0                    py36_0  
importlib_metadata        1.7.0                         0  
iniconfig                 1.0.1                      py_0  
intel-openmp              2020.2                      254  
joblib                    0.16.0                   pypi_0    pypi
jpeg                      9b                   h024ee3a_2  
json-c                    0.13.1               h1bed415_0  
kealib                    1.4.7                hd0c454d_6  
kiwisolver                1.2.0            py36hfd86e86_0  
krb5                      1.16.4               h173b8e3_0  
lcms2                     2.11                 h396b838_0  
ld_impl_linux-64          2.33.1               h53a641e_7  
libboost                  1.67.0               h46d08c1_4  
libcurl                   7.67.0               h20c2e04_0  
libdap4                   3.19.1               h6ec2957_0  
libedit                   3.1.20191231         h14c3975_1  
libevent                  2.1.8                h1ba5d50_1  
libffi                    3.2.1                hd88cf55_4  
libgcc-ng                 9.1.0                hdf63c60_0  
libgdal                   3.0.2                h27ab9cc_0  
libgfortran-ng            7.3.0                hdf63c60_0  
libkml                    1.3.0                h590aaf7_4  
libnetcdf                 4.6.1                h11d0813_2  
libpng                    1.6.37               hbc83047_0  
libpq                     11.2                 h20c2e04_0  
libprotobuf               3.11.2               hd408876_0  
libspatialindex           1.9.3                he6710b0_0  
libspatialite             4.3.0a               h793db0d_0  
libssh2                   1.9.0                h1ba5d50_1  
libstdcxx-ng              9.1.0                hdf63c60_0  
libtiff                   4.1.0                h2733197_0  
libuuid                   1.0.3                h1bed415_2  
libxcb                    1.14                 h7b6447c_0  
libxml2                   2.9.10               he19cac6_1  
lz4-c                     1.8.1.2              h14c3975_0  
mapclassify               2.2.0                    pypi_0    pypi
matplotlib                3.3.1                         0  
matplotlib-base           3.3.1            py36h817c723_0  
mkl                       2020.2                      256  
mkl-service               2.3.0            py36he904b0f_0  
mkl_fft                   1.2.0            py36h23d657b_0  
mkl_random                1.1.1            py36h0573a6f_0  
more-itertools            8.5.0                      py_0  
munch                     2.5.0                      py_0  
ncurses                   6.2                  he6710b0_1  
numpy                     1.19.1           py36hbc911f0_0  
numpy-base                1.19.1           py36hfa32c7d_0  
olefile                   0.46                     py36_0  
openjpeg                  2.3.0                h05c96fa_1  
openssl                   1.1.1h               h7b6447c_0  
packaging                 20.4                       py_0  
pandas                    0.25.3           py36he6710b0_0  
pcre                      8.44                 he6710b0_0  
pillow                    7.2.0            py36hb39fc2d_0  
pip                       20.2.2                   py36_0  
pixman                    0.40.0               h7b6447c_0  
pluggy                    0.13.1                   py36_0  
poppler                   0.65.0               h581218d_1  
poppler-data              0.4.9                         0  
postgresql                11.2                 h20c2e04_0  
proj                      6.2.1                haa6030c_0  
py                        1.9.0                      py_0  
pyarrow                   0.15.1           py36h0573a6f_0  
pygeos                    0.8+7.ged7b3fa           pypi_0    pypi
pyparsing                 2.4.7                      py_0  
pyproj                    2.3.1                    pypi_0    pypi
pyqt                      5.9.2            py36h05f1152_2  
pytest                    6.0.2                    py36_0  
pytest-cov                2.10.1                     py_0  
pytest-forked             1.3.0                      py_0  
pytest-xdist              2.1.0                      py_0  
python                    3.6.10               h191fe78_1  
python-dateutil           2.8.1                      py_0  
pytz                      2020.1                     py_0  
qt                        5.9.7                h5867ecd_1  
re2                       2019.08.01           he6710b0_0  
readline                  7.0                  h7b6447c_5  
rtree                     0.9.4                    py36_1  
scikit-learn              0.23.2                   pypi_0    pypi
scipy                     1.5.2                    pypi_0    pypi
setuptools                49.6.0                   py36_0  
shapely                   1.7.1            py36h98ec03d_0  
sip                       4.19.8           py36hf484d3e_0  
six                       1.15.0                     py_0  
snappy                    1.1.8                he6710b0_0  
sqlalchemy                1.3.19           py36h7b6447c_0  
sqlite                    3.33.0               h62c20be_0  
tbb                       2018.0.5             h6bb024c_0  
threadpoolctl             2.1.0                    pypi_0    pypi
thrift-cpp                0.11.0               h02b749d_3  
tiledb                    1.6.3                h1fb8f14_0  
tk                        8.6.10               hbc83047_0  
toml                      0.10.1                     py_0  
tornado                   6.0.4            py36h7b6447c_1  
uriparser                 0.9.3                he6710b0_1  
wheel                     0.35.1                     py_0  
wrapt                     1.12.1                   pypi_0    pypi
xerces-c                  3.2.3                h780794e_0  
xz                        5.2.5                h7b6447c_0  
zipp                      3.1.0                      py_0  
zlib                      1.2.11               h7b6447c_3  
zstd                      1.3.7                h0b5b093_0  
Setting HAS_PYGEOS=1
```
Env with success:
```
_libgcc_mutex             0.1                        main  
apipkg                    1.5                      py36_0  
arrow-cpp                 0.15.1           py36hfa00764_6  
attrs                     20.2.0                     py_0  
blas                      1.0                         mkl  
boost-cpp                 1.67.0               h14c3975_4  
brotli                    1.0.9                he6710b0_0  
bzip2                     1.0.8                h7b6447c_0  
c-ares                    1.16.1               h7b6447c_0  
ca-certificates           2020.7.22                     0  
cairo                     1.14.12              h8948797_3  
certifi                   2020.6.20                py36_0  
cfitsio                   3.470                hb7c8383_2  
click                     7.1.2                      py_0  
click-plugins             1.1.1                      py_0  
cligj                     0.5.0                    py36_0  
coverage                  5.3              py36h7b6447c_0  
curl                      7.67.0               hbc83047_0  
cycler                    0.10.0                   py36_0  
dbus                      1.13.16              hb2f20db_0  
deprecated                1.2.10                   pypi_0    pypi
descartes                 1.1.0                      py_4  
double-conversion         3.1.5                he6710b0_1  
execnet                   1.7.1                      py_0  
expat                     2.2.9                he6710b0_2  
fiona                     1.8.13.post1     py36hc820daa_0  
fontconfig                2.13.0               h9420a91_0  
freetype                  2.10.2               h5ab3b9f_0  
freexl                    1.0.5                h14c3975_0  
fsspec                    0.8.0                      py_0  
gdal                      3.0.2            py36hbb2a789_0  
geographiclib             1.50                     pypi_0    pypi
geopy                     2.0.0                    pypi_0    pypi
geos                      3.8.0                he6710b0_0  
geotiff                   1.5.1                h21e8280_1  
gflags                    2.2.2                he6710b0_0  
giflib                    5.1.4                h14c3975_1  
glib                      2.63.1               h5a9c865_0  
glog                      0.4.0                he6710b0_0  
grpc-cpp                  1.26.0               hf8bcb03_0  
gst-plugins-base          1.14.0               hbbd80ab_1  
gstreamer                 1.14.0               hb453b48_1  
hdf4                      4.2.13               h3ca952b_2  
hdf5                      1.10.4               hb1b8bf9_0  
icu                       58.2                 he6710b0_3  
importlib-metadata        1.7.0                    py36_0  
importlib_metadata        1.7.0                         0  
iniconfig                 1.0.1                      py_0  
intel-openmp              2020.2                      254  
joblib                    0.16.0                   pypi_0    pypi
jpeg                      9b                   h024ee3a_2  
json-c                    0.13.1               h1bed415_0  
kealib                    1.4.7                hd0c454d_6  
kiwisolver                1.2.0            py36hfd86e86_0  
krb5                      1.16.4               h173b8e3_0  
lcms2                     2.11                 h396b838_0  
ld_impl_linux-64          2.33.1               h53a641e_7  
libboost                  1.67.0               h46d08c1_4  
libcurl                   7.67.0               h20c2e04_0  
libdap4                   3.19.1               h6ec2957_0  
libedit                   3.1.20191231         h14c3975_1  
libevent                  2.1.8                h1ba5d50_1  
libffi                    3.2.1                hd88cf55_4  
libgcc-ng                 9.1.0                hdf63c60_0  
libgdal                   3.0.2                h27ab9cc_0  
libgfortran-ng            7.3.0                hdf63c60_0  
libkml                    1.3.0                h590aaf7_4  
libnetcdf                 4.6.1                h11d0813_2  
libpng                    1.6.37               hbc83047_0  
libpq                     11.2                 h20c2e04_0  
libprotobuf               3.11.2               hd408876_0  
libspatialindex           1.9.3                he6710b0_0  
libspatialite             4.3.0a               h793db0d_0  
libssh2                   1.9.0                h1ba5d50_1  
libstdcxx-ng              9.1.0                hdf63c60_0  
libtiff                   4.1.0                h2733197_0  
libuuid                   1.0.3                h1bed415_2  
libxcb                    1.14                 h7b6447c_0  
libxml2                   2.9.10               he19cac6_1  
lz4-c                     1.8.1.2              h14c3975_0  
mapclassify               2.2.0                    pypi_0    pypi
matplotlib                3.3.1                         0  
matplotlib-base           3.3.1            py36h817c723_0  
mkl                       2020.2                      256  
mkl-service               2.3.0            py36he904b0f_0  
mkl_fft                   1.2.0            py36h23d657b_0  
mkl_random                1.1.1            py36h0573a6f_0  
more-itertools            8.5.0                      py_0  
munch                     2.5.0                      py_0  
ncurses                   6.2                  he6710b0_1  
numpy                     1.19.1           py36hbc911f0_0  
numpy-base                1.19.1           py36hfa32c7d_0  
olefile                   0.46                     py36_0  
openjpeg                  2.3.0                h05c96fa_1  
openssl                   1.1.1h               h7b6447c_0  
packaging                 20.4                       py_0  
pandas                    0.25.3           py36he6710b0_0  
pcre                      8.44                 he6710b0_0  
pillow                    7.2.0            py36hb39fc2d_0  
pip                       20.2.2                   py36_0  
pixman                    0.40.0               h7b6447c_0  
pluggy                    0.13.1                   py36_0  
poppler                   0.65.0               h581218d_1  
poppler-data              0.4.9                         0  
postgresql                11.2                 h20c2e04_0  
proj                      6.2.1                haa6030c_0  
py                        1.9.0                      py_0  
pyarrow                   0.15.1           py36h0573a6f_0  
pygeos                    0.8+4.g249dabc           pypi_0    pypi
pyparsing                 2.4.7                      py_0  
pyproj                    2.3.1                    pypi_0    pypi
pyqt                      5.9.2            py36h05f1152_2  
pytest                    6.0.2                    py36_0  
pytest-cov                2.10.1                     py_0  
pytest-forked             1.3.0                      py_0  
pytest-xdist              2.1.0                      py_0  
python                    3.6.10               h191fe78_1  
python-dateutil           2.8.1                      py_0  
pytz                      2020.1                     py_0  
qt                        5.9.7                h5867ecd_1  
re2                       2019.08.01           he6710b0_0  
readline                  7.0                  h7b6447c_5  
rtree                     0.9.4                    py36_1  
scikit-learn              0.23.2                   pypi_0    pypi
scipy                     1.5.2                    pypi_0    pypi
setuptools                49.6.0                   py36_0  
shapely                   1.7.1            py36h98ec03d_0  
sip                       4.19.8           py36hf484d3e_0  
six                       1.15.0                     py_0  
snappy                    1.1.8                he6710b0_0  
sqlalchemy                1.3.19           py36h7b6447c_0  
sqlite                    3.33.0               h62c20be_0  
tbb                       2018.0.5             h6bb024c_0  
threadpoolctl             2.1.0                    pypi_0    pypi
thrift-cpp                0.11.0               h02b749d_3  
tiledb                    1.6.3                h1fb8f14_0  
tk                        8.6.10               hbc83047_0  
toml                      0.10.1                     py_0  
tornado                   6.0.4            py36h7b6447c_1  
uriparser                 0.9.3                he6710b0_1  
wheel                     0.35.1                     py_0  
wrapt                     1.12.1                   pypi_0    pypi
xerces-c                  3.2.3                h780794e_0  
xz                        5.2.5                h7b6447c_0  
zipp                      3.1.0                      py_0  
zlib                      1.2.11               h7b6447c_3  
zstd                      1.3.7                h0b5b093_0  
Setting HAS_PYGEOS=1
```

The only change that jumped out to me was `pygeos`. Given that most of the failures explicitly relate to null geom types, and https://github.com/pygeos/pygeos/pull/199/files fixed a bug related to null geometries and was merged 12 hours, guessing there might be some connection there, but I haven't looked any deeper than that so it may well be a red herring.

File: geopandas/geodataframe.py [hunk 1]
```diff
@@ -512,6 +512,9 @@ def to_json(self, na="null", show_bbox=False, **kwargs):
             See below.
         show_bbox : bool, optional, default: False
             Include bbox (bounds) in the geojson
+        drop_id : bool, optional
+            Whether to retain the id. Default is False, but may want True
```
Comment (martinfleis) [inline geopandas/geodataframe.py:516]: It is not entirely clear that the `id` refers to the actual index of gdf. Can we expand the description a bit to make sure everyone knows that `drop_id` controls whether to preserve index as `id` property of json or not? Same applies to other docstrings below.

**Review by martinfleis — COMMENTED** [review]
Thanks! Just one comment regarding docstring.

The CI will fail again, but the fix is in #1644.

File: geopandas/geodataframe.py (hunk 1 above)
Comment (jdmcbr) [reply geopandas/geodataframe.py:516]: Thanks, that's a good point.

File: geopandas/geodataframe.py [hunk 2]
```diff
@@ -703,6 +714,7 @@ def iterfeatures(self, na="null", show_bbox=False):
         >>> feature
         {'id': '0', 'type': 'Feature', 'properties': {'col1': 'name1'}, 'geometry': {\
 'type': 'Point', 'coordinates': (1.0, 2.0)}}
+>>>>>>> master
```
Comment (martinfleis) [inline geopandas/geodataframe.py:717]: A forgotten line from git conflict.

Comment (jdmcbr) [reply geopandas/geodataframe.py]: :facepalm:

**Comment by jdmcbr** [comment]
@martinfleis There's a new failure on one of the envs after merging master in again. There are a handful of doctests that are now failing because the id is inserted last instead of being inserted first, so the order in which the members of the dict are printed has changed. Incidentally, I'm kind of confused on why this is only failing on one env—once I realized what was going on, I remembered that python 3.7 has actually preserved the order of keys in a dict, so it seems like more of the envs should be failing. Anyway, do you have thoughts on either updating the code so the `'id'` is added first (which seems a little ugly to me) or updating the doctest to be less natural in terms of real world usage (which also seems a little ugly to me) or running relevant doctest functions with `drop_id=True` so the ordering doesn't matter (which feels least bad but I still don't love)?

**Comment by martinfleis** [comment]
> I'm kind of confused on why this is only failing on one env

Because docstrings are tested in that one environment only. There may slight differences in output between the versions so I turned it on in a single one only in #1659 

I would not do `drop_id=True`, we would have to explain in those doctstring why we do it.

We can either update docstrings according to the actual behaviour, i.e. move `id` to the last position, or change the code. It may be actually better to fix this in the code, because you are right that json normally comes with id first.

Something like this is not ideal but I don't think there's a better option if we want to preserve order.

```python
if drop_id:
    feature = {
          "type": "Feature",
          "properties": properties_items,
          "geometry": mapping(geom) if geom else None,
      }
else:
    feature = {
          "id": str(ids[i]),
          "type": "Feature",
          "properties": properties_items,
          "geometry": mapping(geom) if geom else None,
      }
```

**Comment by jdmcbr** [comment]
> Because docstrings are tested in that one environment only. There may slight differences in output between the versions so I turned it on in a single one only in #1659

Aha, I'd missed that one.

Anyway, went ahead and updated so `id` is the first parameter set. Better to have that first since that's typical with JSON, as you note.

**Review by martinfleis — APPROVED** [review]
Thanks! It is looking good now.

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
One very late thought: we could also call this `include_id` (or `write_id`) to be affirmative in the default action (but no strong opinion here)

**Comment by jdmcbr** [comment]
@jorisvandenbossche I don't have an especially strong opinion there. I found `drop_id` somehow slightly more intuitive, in that it asks for affirmation to change the current default behavior. But I'm happy to change it if others think the verb in the keyword argument should be an affirmative one.

**Comment by jorisvandenbossche** [comment]
That's true, I am fine with `drop_id` as well!

**Comment by martinfleis** [comment]
I am fine with `drop_id` as well. Merged! Thanks @jdmcbr!

**Comment by kiranzo** [comment]
Hey, guys @jdmcbr @martinfleis , just bringing to your attention that while you fixed this issue for GeoDataFrame, id field is still obligatory for GeoSeries, there is no option to remove it. You might want to add it for GeoSeries as well.

My version of geopandas is 0.14.2.

**Pull request merged**